cql3/restrictions/statement_restrictions.cc

```diff
@@ -969,5 +975,229 @@ bool single_column_restriction::LIKE::is_satisfied_by(bytes_view data, const que
     return (*_matcher)(data);
 }
 
+namespace wip {
+
+namespace {
+
+using cql3::selection::selection;
+
+/// Returns cdef's value from the fetched data.
+bytes_opt get_value(const column_definition* cdef, const selection& selection,
+                    const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+                    const std::vector<bytes_opt>& other_columns) {
+    switch (cdef->kind) {
+    case column_kind::partition_key:
+        return partition_key[cdef->id];
+    case column_kind::clustering_key:
+        return clustering_key[cdef->id];
+    case column_kind::static_column:
+    case column_kind::regular_column:
+        return other_columns[selection.index_of(*cdef)];
+    default:
+        throw exceptions::unsupported_operation_exception("Unknown column kind");
+    }
 }
+
+/// True iff columns' values equal t.
+bool equal(::shared_ptr<term> t, const idents& columns, const selection& selection,
+           const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+           const std::vector<bytes_opt>& other_columns, const query_options& options) {
+    if (auto multi = dynamic_pointer_cast<multi_item_terminal>(t)) {
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
```

Comment:
How can this happen? We should have a prepare-time check.

Reply:
We do have that check, this is just defensive programming.  I can remove it, if you wish, or enclose it in `__builtin_expect`?

Reply:
throw already implies expect(false).

Ideally we would have a way to prove that this can't happen. By splitting this into one function per case, and having prepare() select the correct function and store it in the prepared statement.

Reply:
Agreed, but that's far in the future.  This is two steps removed from prepare().  I added __builtin_expect in the next version.